coda-src/venus/9pfs.cc

```diff
@@ -1086,22 +1095,66 @@ int plan9server::recv_clunk(unsigned char *buf, size_t len, uint16_t tag)
 int plan9server::recv_remove(unsigned char *buf, size_t len, uint16_t tag)
 {
     uint32_t fid;
+    int rc;
 
     if (unpack_le32(&buf, &len, &fid))
         return -1;
 
     DEBUG("9pfs: Tremove[%x] fid %u\n", tag, fid);
 
-#if 0
+    struct fidmap *fm = find_fid(fid);
+    if (!fm)
+        return send_error(tag, "fid unknown or out of range");
+
+    /* find the filename */
+    char name[NAME_MAX] = "???";
+    /* Coda supports hardlinks so there may be multiple valid names for the
+     * same file. 9pfs doesn't handle hardlinks so each file can maintain
+     * a unique name. Coda does track the last name used to lookup the object,
+     * so return that. */
+    fsobj *f = FSDB->Find(&fm->cnode.c_fid);
+        if (f) f->GetPath(name, PATH_COMPONENT);
+
+    /* Find the parent directory cnode
+     * We need to move back up the tree to get past the possible P9 fids
```

Comment:
I think we can also use `fsobj *f->pfso`, or `Find(f->pfid)` when `pfso == NULL`. That we traverse the namespace in Coda instead of iterating through the copy of the namespace we've constructed in the form of the fidmap.